database.cc

```diff
@@ -1441,8 +1441,8 @@ database::query(schema_ptr s, const query::read_command& cmd, query::result_opti
     if (querier_opt) {
         co_await querier_opt->close();
     }
-    if(ex) {
-        std::rethrow_exception(std::move(ex));
+    if (ex) {
+        co_return coroutine::exception(ex);
```

Comment:
Why not std::move(ex) here?
I understand that it's out of fashion to do "return std::move(ex)". But here we don't return ex directly, we run a function coroutine::exception() on it. Doesn't this function call need to know that it doesn't need to copy ex?
You used the same idiom in many places so I won't ask it on all the relevant places, just once here.

Reply:
`coroutine::exception`'s constructor takes this parameter by value anyway in order to cut such dilemmas short :) `std::exception_ptr`'s size is not defined by the standard, but actually it's 1 pointer in clang++ and g++ and reportedly 2 pointers in vc++, so it's as cheap to copy as any other pointer.

Reply:
Please correct me if I'm wrong, but std::exception_ptr is reference-counted like std::shared_ptr, so copying it involves an **atomic** counter increment and decrement, so it's significantly worse than copying a pointer. This is why we std::move()ed exception_ptr when we could.
Perhaps the compiler can optimize away these copies, but I'm not sure at all that it is allowed to optimize away atomic operations.

Reply:
Ok, let me verify what happens, that's a good point.

Reply:
Ouch, you're absolutely right, the copy constructor operates on atomics (ref: https://code.woboq.org/gcc/libstdc++-v3/libsupc++/eh_ptr.cc.html#102) and experiments show that the copy is not really optimized out... In that case I think I'll start with changing the constructor declaration in Seastar, since the patch is still very fresh and it's likely that nobody relies on it yet.

Reply:
Ok, I sent a few Seastar patches to address the problem. Let's put this pull request on hold until the issue is resolved in Seastar